### PR TITLE
Parent node is set for bound StateNodes

### PR DESCRIPTION
**Pull request by caalador**
This change is [Reviewable](https://reviewable.io/reviews/vaadin/flow/3183)

### CODE REVIEW

**Comment by denis-anisimov** [comment]
Reviewed 2 of 4 files at r1.
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0iZ1lfTlUTc3kTXsD0:b-o8iaiw) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
> shadowRootNode.setParent(context.node);

I would suggest to avoid this way of implementation.
- It relies on specific strategy. We have many strategies. Some of them do their own binding. (Though most of those strategies should be just removed. But still `TextBindingStrategy` will be kept).
- There are many places where the parent needs to be set. The logic becomes too fragile.

I suggest to have only one place where the parent is set: `MapProperty` .
I'm not sure about setter ( whether it's always called directly or there is magic which sets the value in  the property).
But we can use getter at least ( I would prefer setter since it will be called either once or significantly  less often).

All `StateNode` are in some `NodeMap`s. There is no another way to get `StateNode`.
So we can just this code inside `MapProperty`:

```
if ( value instanceof StateNode ){
  StateNode node = (StateNode) value;
  if ( node.getParent() == null){
     node.setParent(getMap().getNode());
  }
}
```

Better if it's in the setter. But can be in the getter ( `getValue()`/`setValue()` )

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by caalador** [comment]
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0iclX3GwiMiagGpKON:b-6jps4t) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> shadowRootNode.setParent(context.node);

I would suggest to avoid this way of implementation.
- It relies on specific strategy. We have many strategies. Some of them do their own binding. (Though most of those strategies should be just removed. But still `TextBindingStrategy` will be kept).
- There are many places where the parent needs to be set. The logic becomes too fragile.

I suggest to have only one place where the parent is set: `MapProperty` .
I'm not sure about setter ( whether it's always called directly or there is magic which sets the value in  the property).
But we can use getter at least ( I would prefer setter since it will be called either once or significantly  less often).

All `StateNode` are in some `NodeMap`s. There is no another way to get `StateNode`.
So we can just this code inside `MapProperty`:

```
if ( value instanceof StateNode ){
  StateNode node = (StateNode) value;
  if ( node.getParent() == null){
     node.setParent(getMap().getNode());
  }
}
```

Better if it's in the setter. But can be in the getter ( `getValue()`/`setValue()` )
</blockquote></details>

There is just the problem that all of the StateNodes that come in there are nodes with no features and they are not the actual component StateNodes.

`TextBindingStrategy` doesn't have child nodes so that is not needed as it doesn't allow child nodes. 

I'll try to find where we would know the correct information at, but...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by denis-anisimov** [comment]
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0iicPoBe1PthxtXQXn:bxyi06p) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

There is just the problem that all of the StateNodes that come in there are nodes with no features and they are not the actual component StateNodes.

`TextBindingStrategy` doesn't have child nodes so that is not needed as it doesn't allow child nodes. 

I'll try to find where we would know the correct information at, but...
</blockquote></details>

We should not care whether they are component `StateNode`s actually.

In reality it's totally fine to have non-component `StateNode`.
E.g. the `StateNode`  may represent the model bean.

The main usecase here is: check whether an ascendant is "not visible".
In the latter case we should not update property value for such node ( at least. Also may be not send RPC).
This usecase is true also for `StateNode` which represents model bean.
We should not update such bean.
So it's totally fine when `StateNode`  is not a component node.

This feature is just a replication of the `StateNode` hierarchy from the server side.
You shouldn't care about nature of the `StateNode`.  
If the node in the NodeMap (or NodeList , it also should be checked) then this map or list has a "parent".
This parent should be set as a parent for any node in the map/list.
That's it. It's just a pure hierarchical question without any additional semantic.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by denis-anisimov** [comment]
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0ilMozUtFKx7YwW5Ib:byerrys) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

We should not care whether they are component `StateNode`s actually.

In reality it's totally fine to have non-component `StateNode`.
E.g. the `StateNode`  may represent the model bean.

The main usecase here is: check whether an ascendant is "not visible".
In the latter case we should not update property value for such node ( at least. Also may be not send RPC).
This usecase is true also for `StateNode` which represents model bean.
We should not update such bean.
So it's totally fine when `StateNode`  is not a component node.

This feature is just a replication of the `StateNode` hierarchy from the server side.
You shouldn't care about nature of the `StateNode`.  
If the node in the NodeMap (or NodeList , it also should be checked) then this map or list has a "parent".
This parent should be set as a parent for any node in the map/list.
That's it. It's just a pure hierarchical question without any additional semantic.
</blockquote></details>

So in short: every `StateNode` except the root should have a parent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by caalador** [comment]
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0il-EoEkP6EpAapcbb:b-m08kxg) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

So in short: every `StateNode` except the root should have a parent.
</blockquote></details>

yes, But with the MapProperty none of the StateNodes in the StateTree node list have a parent StateNode.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by denis-anisimov** [comment]
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0imtzg2uTJ7nCBmk5_:bwktn2t) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

yes, But with the MapProperty none of the StateNodes in the StateTree node list have a parent StateNode.
</blockquote></details>

Sorry, I'm confused.

Do you mean that the `MapProperty.getMap().getNode()` is `null` ?
Or which parent you mean? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by denis-anisimov** [comment]
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0incZ9Vq092zqPTkR1:b-ijw8u4) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

Sorry, I'm confused.

Do you mean that the `MapProperty.getMap().getNode()` is `null` ?
Or which parent you mean? 
</blockquote></details>

So this doesn't work ?

```

private void updateValue(Object value, boolean hasValue) {
        Object oldValue = this.value;

        this.hasValue = hasValue;
        this.value = value;

        if ( value instanceof StateNode ){
  StateNode node = (StateNode) value;
  if ( node.getParent() == null){
     node.setParent(getMap().getNode());
  }
}

        eventRouter
                .fireEvent(new MapPropertyChangeEvent(this, oldValue, value));
    }
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by caalador** [comment]
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0io4xoIBUnrh-wbvzM:b7s743i) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

So this doesn't work ?

```

private void updateValue(Object value, boolean hasValue) {
        Object oldValue = this.value;

        this.hasValue = hasValue;
        this.value = value;

        if ( value instanceof StateNode ){
  StateNode node = (StateNode) value;
  if ( node.getParent() == null){
     node.setParent(getMap().getNode());
  }
}

        eventRouter
                .fireEvent(new MapPropertyChangeEvent(this, oldValue, value));
    }
```
</blockquote></details>

No. It sets parent StateNodes to some StateNodes, but when actually checking the StateTree idToNode map none of the StateNodes there have a parent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by denis-anisimov** [comment]
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 317 at r1](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0iZ1lfTlUTc3kTXsD-:-L0ip5J9h8jzgtw-XggG:b-xyx7ru) ([raw file](https://github.com/vaadin/flow/blob/66cdf7b2532a6cfa583ccc4e4da7dfee78a7d3b1/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L317)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

No. It sets parent StateNodes to some StateNodes, but when actually checking the StateTree idToNode map none of the StateNodes there have a parent.
</blockquote></details>

Weird.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by caalador** [comment]
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0mhVIF0HdFgvVlyOUl:-L0mhVIF0HdFgvVlyOUm:b-of1m29) (no related file):*
What about this route?
Then we don't care all too much about what or where.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by denis-anisimov** [comment]
Reviewed 4 of 5 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0mhVIF0HdFgvVlyOUl:-L0mxs7s7hS3GgoqjSIR:by2ujsl) (no related file):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

What about this route?
Then we don't care all too much about what or where.
</blockquote></details>

:+1: 

But do you still need `GwtBasicElementBinderTest` ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*

**Comment by caalador** [comment]
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/3183#-L0mhVIF0HdFgvVlyOUl:-L0nCj7xDoGgrlk1zHX2:bca3265) (no related file):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

:+1: 

But do you still need `GwtBasicElementBinderTest` ?
</blockquote></details>

No. It was supposed to be reverted, but I missed that part it seems.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3183)*
